runtime/autoload/provider/pythonx.vim

```diff
@@ -22,96 +24,122 @@ function! provider#pythonx#Require(host) abort
+  for item in split(path, path_sep)
+    if !has_key(already_seen, item)
+      let already_seen[item] = v:true
+      call add(path_list, item)
+    endif
+  endfor
+
+  return join(path_list, ',')
 endfunction
 
-function! s:get_python_candidates(major_version) abort
-  return {
-        \ 2: ['python2', 'python2.7', 'python2.6', 'python'],
-        \ 3: ['python3', 'python3.9', 'python3.8', 'python3.7', 'python3.6', 'python3.5',
-        \     'python3.4', 'python3.3', 'python']
-        \ }[a:major_version]
+" Returns a list of all Python executables found on path. a:convert specifies
+" whether the path to search needs to be converted from a host-specific
+" separator to the comma-separated format expected by globpath().
+function! provider#pythonx#GetPythonCandidates(major_version, path, convert) abort
+  let path = a:convert ? s:to_comma_separated_path(a:path) : a:path
+  let starts_with_python = globpath(path, 'python*', v:true, v:true)
+  let ext_pat = has('win32') ? '(\\.exe)?' : ''
+  let matches_version = printf('v:val =~# "\\v[\\/]python(%s)?(\\.[0-9]+)?%s$"', a:major_version, ext_pat)
+  return filter(starts_with_python, matches_version)
 endfunction
 
-" Returns [path_to_python_executable, error_message]
+" Returns [path_to_python_executable, python_version, error_messages]
 function! provider#pythonx#Detect(major_version) abort
   return provider#pythonx#DetectByModule('neovim', a:major_version)
 endfunction
 
-" Returns [path_to_python_executable, error_message]
+" Returns [path_to_python_executable, python_version, error_messages]
 function! provider#pythonx#DetectByModule(module, major_version) abort
-  let python_exe = s:get_python_executable_from_host_var(a:major_version)
+  let host_prog = 'python'.(a:major_version == 3 ? '3' : '').'_host_prog'
+  let python_exe = get(g:, host_prog, '')
+  let errors = []
 
   if !empty(python_exe)
-    return [exepath(expand(python_exe)), '']
+    let candidates = [exepath(expand(python_exe))]
+    call add(errors, 'The g:'.host_prog.' you set cannot be used.')
+  else
+    let candidates = provider#pythonx#GetPythonCandidates(a:major_version, $PATH, v:true)
   endif
 
-  let candidates = s:get_python_candidates(a:major_version)
-  let errors = []
+  if empty(candidates)
+    call add(errors, 'No candidates for a Python '.a:major_version.' executable found on $PATH.')
+  endif
 
   for exe in candidates
-    let [result, error] = provider#pythonx#CheckForModule(exe, a:module, a:major_version)
+    let [result, python_version, error] = provider#pythonx#CheckForModule(exe, a:module, a:major_version)
     if result
-      return [exe, error]
+      return [exe, python_version, error]
     endif
     " Accumulate errors in case we don't find any suitable Python executable.
     call add(errors, error)
   endfor
 
   " No suitable Python executable found.
-  return ['', 'provider/pythonx: Could not load Python '.a:major_version.":\n".join(errors, "\n")]
+  return ['', '', errors]
```

Comment:
It's better to return a list of errors, the health check formats it prettily. And the "provider/pythonx: Could not load Python etc." message feels a bit redundant in the context that it's shown in.